[LoongArch64] Sync with PR#107282: Transfer arguments between calling conventions in shuffling thunks.

LuckyXu-HF opened the pull request:
[LoongArch64] Sync with [PR#107282](https://github.com/dotnet/runtime/pull/107282): Transfer arguments between calling conventions in shuffling thunks.
* Set 'varDsc->lvIsMultiRegArg = true' when passing struct with two regs in 'Compiler::lvaInitUserArgs()' to avoid changing passing struct with two regs into one regs in Prolog.

LuckyXu-HF committed:
[LoongArch64] Sync with PR#107282: Transfer arguments between calling conventions in shuffling thunks.
* Set `varDsc->lvIsMultiRegArg = true` when passing struct with two regs in Compiler::lvaInitUserArgs()
to avoid changing passing struct with two regs into one regs in Prolog.

LuckyXu-HF commented:
@tomeksowi Thanks very much for your notification and information.
@shushanhf @jkotas could you please review this PR? Thanks.

shushanhf reviewed (approved):
LGTM
Thanks

tomeksowi commented on src/coreclr/vm/loongarch64/stubs.cpp:
```diff
@@ -1277,109 +1305,43 @@ VOID StubLinkerCPU::EmitShuffleThunk(ShuffleEntry *pShuffleEntryArray)
-                    delay_index[index++] = i;
-                    continue;
-                }
-                EmitLoadFloatRegImm(FloatReg((pEntry->dstofs & ShuffleEntry::OFSREGMASK)), RegSp, pEntry->srcofs * sizeof(void*));
```
I think that was the last use, you can remove `Emit*Float*` methods and FloatReg itself.

LuckyXu-HF replied:
Yes you are right, thank you.

tomeksowi commented on src/coreclr/jit/lclvars.cpp:
```diff
@@ -1021,6 +1021,10 @@ void Compiler::lvaInitUserArgs(InitVarDscInfo* varDscInfo, unsigned skipArgs, un
                             genMapRegArgNumToRegNum(secondAllocatedRegArgNum, argRegTypeInStruct2, info.compCallConv));
+
+#ifdef TARGET_LOONGARCH64
+                        varDsc->lvIsMultiRegArg = true;
```
Which case does that fix? I'm wondering why it worked for RISC-V and not for LoongArch, both calling conventions are very similar.

LuckyXu-HF replied:
Yeah, I'm also very curious about this. The only case I meet about this was mentioned in https://github.com/dotnet/runtime/pull/107282#issuecomment-2416635910 .

It’s about the codegen for the Prolog of `MethodHash=88c18db0` with `DOTNET_TieredCompilation=0` (or `MethodHash=88c18db0 Tier1` with `DOTNET_TieredCompilation=1`) for `EmptyStructs.sh` under Debug mode on LA64.
* The corresponding JIT function is EmptyStructs.cs:1726 `Assert.Equal(FloatEmptyShort.Get(), stack0_to_fa6_a7);`

LA64 codegen will try to merge the first struct from two registers (`a0 & f1`) into one register (`s1)`, this does not match LA64’s ABI. So then I will meet the assertion failure of codegencommon.cpp:3549 `noway_assert(!"Insertion into register is not supported");` about this.

As we did use multi reg to pass struct here, I think we should set the `varDsc->lvIsMultiRegArg = true;` like arm64 and amd64 to solve this problem.

I don't have a RISCV64 environment, so it’s hard to track RV64’s behavior of this case. Based on this change LA64’s Prolog about this is liking:
```shell
G_M29263_IG01:  ;; offset=0x0000
  02FEC063  addi.d          sp, sp, -80
  29C04076  st.d            fp, sp, 16 
  29C06061  st.d            ra, sp, 24 
  29C08077  st.d            s0, sp, 32 
  29C0A078  st.d            s1, sp, 40 
  29C0C079  st.d            s2, sp, 48 
  29C0E07A  st.d            s3, sp, 56 
  29C1007B  st.d            s4, sp, 64 
  02C04076  addi.d          fp, sp, 16 
  02C1406A  addi.d          a6, sp, 80 
  29C0E2CA  st.d            a6, fp, 56 
  2B7FE2C0  fst.s           f0, fp, -8 
  297FFAC5  st.h            a1, fp, -2 
  2B7FC2C1  fst.s           f1, fp, -16
  297FDAC7  st.h            a3, fp, -10
  03800097  mov             s0, a0 
  038000D8  mov             s1, a2                
  0380011A  mov             s3, a4 
  03800139  mov             s2, a5 
```

LuckyXu-HF replied:
BTW, I think one more place maybe hit a risk in the above case in morph.cpp:2961-2963:
* I think the `lclNum` get from `lclNode->GetLclNum()` is not completely equivalent with the `varDsc` index in `lvaGetDesc`:
```csharp
####morph.cpp:
...
...
 2959     if (argNode->OperIs(GT_LCL_VAR))
 2960     {
 2961         GenTreeLclVarCommon* lclNode = argNode->AsLclVarCommon();                                                                                                                       
 2962         unsigned             lclNum  = lclNode->GetLclNum();
 2963         LclVarDsc*           varDsc  = lvaGetDesc(lclNum);
 2964 
 2965         varDsc->lvIsMultiRegArg = true;
...
...
```
eg: the `JitHashDump=88c18db0` with `DOTNET_TieredCompilation=0` will hit here both on LA64 and linux-arm64:
* We can see the `343.CALL` liking the following, and the lclNum=3 here:
```
46202 Morphing BB15
46203 Assertions in: #NA
46204 fgMorphTree BB15, STMT00079 (before)
46205                [000346] DAC-G------                         *  STORE_LCL_VAR int    V10 loc4         
46206                [000345] --C-G------                         \--*  EQ        int   
46207                [000343] --C-G------                            +--*  CALLV stub int    System.IComparable`1[Program+FloatEmptyShort]:CompareTo(Program+FloatEmptyShort):int:this
46208                [000341] ----------- this                       |  +--*  LCL_VAR   ref    V08 loc2          (last use)
46209                [000342] ----------- arg1                       |  \--*  LCL_VAR   struct<Program+FloatEmptyShort, 8> V03 arg3         
46210                [000344] -----------                            \--*  CNS_INT   int    0
46211 Initializing arg info for 343.CALL:
46212 Argument 0 ABI info: [00..08) reg a0
46213 Argument 1 ABI info: [00..08) reg t8
46214 Argument 2 ABI info: 2 segments  <---@@@@
46215   [0] [00..04) reg f0
46216   [1] [06..08) reg a1
46217 Args for call [000343] CALL after AddFinalArgsAndDetermineABIInfo:
46218 CallArg[[000341].LCL_VAR ref (By value), 1 reg: a0, wellKnown[ThisPointer]]
46219 CallArg[[000442].CNS_INT long (By value), 1 reg: t8, wellKnown[VirtualStubCell]]
46220 CallArg[[000342].LCL_VAR struct (By value), 2 regs: f0 a1]
```

But actually the `LCL_VAR struct` index is 1 here in `lvaGetDesc`,  this will cause a mismatch of the ` morph.cpp: 2963         LclVarDsc*           varDsc  = lvaGetDesc(lclNum);
` , the `varDsc` is not actually what we want here.
```
; Initial local variable assignments
;
;  V00 this              ref  this class-hnd <Xunit.Sdk.AssertEqualityComparer`1[Program+FloatEmptyShort]>
;  V01 arg1           struct <Program+FloatEmptyShort, 8> <Program+FloatEmptyShort>      <-----@@@@
;  V02 arg2              ref  class-hnd <Xunit.Sdk.CollectionTracker>
;  V03 arg3           struct <Program+FloatEmptyShort, 8> <Program+FloatEmptyShort>
;  V04 arg4              ref  class-hnd <Xunit.Sdk.CollectionTracker>
;  V05 arg5            byref 
;  V06 loc0              ref  class-hnd <System.IEquatable`1[Program+FloatEmptyShort]>
;  V07 loc1              ref  class-hnd <System.Collections.IStructuralEquatable>
;  V08 loc2              ref  class-hnd <System.IComparable`1[Program+FloatEmptyShort]>
;  V09 loc3              ref  class-hnd <System.IComparable>
;  V10 loc4            ubyte 
;  V11 loc5              ref  class-hnd <System.Object>
```

LuckyXu-HF replied:
@tomeksowi could you please share some dump information for `JitHashDump=88c18db0` with `DOTNET_TieredCompilation=0` for `343.CALL` if possible? Perhaps this can help to find out some clue for the differences between LA64 and RV64? Thank you.

tomeksowi replied:
> As we did use multi reg to pass struct here, I think we should set the `varDsc->lvIsMultiRegArg = true;` like arm64 and amd64 to solve this problem.

Yes, I think it also shouldn't break anything for RISC-V.

LuckyXu-HF replied:
> > As we did use multi reg to pass struct here, I think we should set the `varDsc->lvIsMultiRegArg = true;` like arm64 and amd64 to solve this problem.
> 
> Yes, I think it also shouldn't break anything for RISC-V.

Does RV64 need to set `varDsc->lvIsMultiRegArg = true` here?
If needed I will delete the `ifdef TARGET_LOONGARCH64`.

tomeksowi replied:
> You can set a breakpoint at `morph.cpp:2810 if (call->gtTreeID == 357)` for `MethodHash=88c18db0` , then set a bp at `morph.cpp:2962` , I think RV64 will also hit the mismatch problem here according to the dump log. At least LA64 and linux-arm64 will hit this mismatch.

Yes, the registers on varDsc are f11 and a3, that looks wrong.

> Does RV64 need to set `varDsc->lvIsMultiRegArg = true` here?
> If needed I will delete the `ifdef TARGET_LOONGARCH64`.

Looks like it, thank you!

LuckyXu-HF replied:
> > You can set a breakpoint at `morph.cpp:2810 if (call->gtTreeID == 357)` for `MethodHash=88c18db0` , then set a bp at `morph.cpp:2962` , I think RV64 will also hit the mismatch problem here according to the dump log. At least LA64 and linux-arm64 will hit this mismatch.
> 
> Yes, the registers on varDsc are f11 and a3, that looks wrong.

Hi @jakobbotsch , could you please help to give us some suggestions about this? linux-arm64 can also hit this: https://github.com/dotnet/runtime/pull/109102#discussion_r1810623051 .
In this special case the arg1 and arg3 are both struct in `lvaGetDesc` with similar members, can we get a correct `lclNum` for the index of `arg1 struct` in `lvaGetDesc` ? Thank you!

> 
> > Does RV64 need to set `varDsc->lvIsMultiRegArg = true` here?
> > If needed I will delete the `ifdef TARGET_LOONGARCH64`.
> 
> Looks like it, thank you!

Updated.

jakobbotsch replied:
There are multi-reg structs used both as arguments and as parameters in this function. For the argument case, they are marked so that we can block enregistering their locals here:
https://github.com/dotnet/runtime/blob/edaa25e6942f3e8d9e2f078d99f34ed015658fbe/src/coreclr/jit/lclvars.cpp#L4038-L4042

For the parameter case, they are marked so that we can block promotion of their fields here:
https://github.com/dotnet/runtime/blob/edaa25e6942f3e8d9e2f078d99f34ed015658fbe/src/coreclr/jit/lclvars.cpp#L2528-L2600

Both these blockers are compensating for things unsupported by the backend. I don't see any mismatches, simply multiple places that things in this function needs to be marked this way (the parameter in `lvaInitUserArgs`, the argument local used in the call in morph).

LuckyXu-HF replied:
> Both these blockers are compensating for things unsupported by the backend. I don't see any mismatches, simply multiple places that things in this function needs to be marked this way (the parameter in `lvaInitUserArgs`, the argument local used in the call in morph).

Thank you, I totally agree with these blockers are indeed effective for the backend.

But during my debugging I found that the `lclNum` get from `lclNode->GetLclNum()` is not completely equivalent with the `varDsc index` in `lvaGetDesc` here:
https://github.com/dotnet/runtime/blob/edaa25e6942f3e8d9e2f078d99f34ed015658fbe/src/coreclr/jit/morph.cpp#L2958-L2962

The eg is LA64:https://github.com/dotnet/runtime/pull/109102#discussion_r1810623051 and RV64:https://github.com/dotnet/runtime/pull/109102#discussion_r1810772288 , and linux-arm64 can also hit this case.
In this case, the `lclNum` get from `lclNode->GetLclNum()` is `3`, but actually what we are dealing with in this node is arg1, and the arg1 is actually in `lvaGetDesc(1)`. So if we take `varDsc` with `lvaGetDesc(lclNum); `here, we will get a wrong `varDsc` which actually represent the arg3, it's mismatch.

LuckyXu-HF replied:
This is a special case what I found during debugging, but currently the backend is able to handle this. I didn't find `morph.cpp:2961-2962` affecting the PASSABLE for any other testcases temporarily now.

jakobbotsch replied:
The value being passed as the argument is `V03` according to your jitdump:
```
46207                [000343] --C-G------                            +--*  CALLV stub int    System.IComparable`1[Program+FloatEmptyShort]:CompareTo(Program+FloatEmptyShort):int:this
46208                [000341] ----------- this                       |  +--*  LCL_VAR   ref    V08 loc2          (last use)
46209                [000342] ----------- arg1                       |  \--*  LCL_VAR   struct<Program+FloatEmptyShort, 8> V03 arg3  // argument passed is value of V03
```

The code looks up the local for the value that is being passed as the argument when it is a `LCL_VAR` node. So I don't see the mismatch you are talking about.

LuckyXu-HF replied:
I'm sorry for not write it clearly, please let me try to describe this more:

The distribution of the arguments in `lvaGetDesc` in morph.cpp:2962 is the following, arg1 and arg3 are both struct with similar members, arg1 will pass by `f0 & a1`, arg3 will pass by `f1 & a3`.
```
; Final local variable assignments
;
;  V00 this         [V00,T00] (  6,  4   )     ref  ->   s0         this class-hnd single-def <Xunit.Sdk.AssertEqualityComparer`1[Program+FloatEmptyShort]>
;  V01 arg1         [V01,T01] (  5,  3.50)  struct ( 8) [fp-0x08]  do-not-enreg[SA] multireg-arg single-def <Program+FloatEmptyShort>
;  V02 arg2         [V02,T03] (  4,  3.50)     ref  ->   s1         class-hnd single-def <Xunit.Sdk.CollectionTracker>
;  V03 arg3         [V03,T02] (  5,  3.50)  struct ( 8) [fp-0x10]  do-not-enreg[SA] multireg-arg single-def <Program+FloatEmptyShort>
;  V04 arg4         [V04,T05] (  4,  3   )     ref  ->   s3         class-hnd single-def <Xunit.Sdk.CollectionTracker>
;  V05 arg5         [V05,T04] (  4,  3.50)   byref  ->   s2         single-def
;* V06 loc0         [V06    ] (  0,  0   )     ref  ->  zero-ref    class-hnd single-def <System.IEquatable`1[Program+FloatEmptyShort]>
;* V07 loc1         [V07    ] (  0,  0   )     ref  ->  zero-ref    class-hnd single-def <System.Collections.IStructuralEquatable>
;* V08 loc2         [V08    ] (  0,  0   )     ref  ->  zero-ref    class-hnd single-def <System.IComparable`1[Program+FloatEmptyShort]>
...
...
```
And the `343.CALL` here with `Compiler::fgMorphMultiregStructArg(CallArg* arg)` is actually dealing with the `arg1(f0 & a1)` above, we can see it in both jitdump log and gdb debug at breakpoint `morph.cpp:2962:`
```
(gdb) p arg->AbiInfo.GetRegNum(0)
$4 = REG_F0
(gdb) p arg->AbiInfo.GetRegNum(1)
$5 = REG_A1

(gdb) p lvaGetDesc(1)._lvArgReg
$8 = 32 ' '  (REG_F0)
(gdb) p lvaGetDesc(1)._lvOtherArgReg
$9 = 5 '\005' (REG_A1)

(gdb) p lvaGetDesc(3)._lvArgReg
$10 = 33 '!'  (REG_F1)
(gdb) p lvaGetDesc(3)._lvOtherArgReg
$11 = 7 '\a' (REG_A3)
```

The `lclNum` get from `lclNode->GetLclNum()` is 3 due to this 343.CALL, and this `CallArg* arg` actually is arg1’s node, the index should be 1 in `lvaGetDesc`. So I think this is a mismatch.
```
Morphing BB15
Assertions in: #NA
fgMorphTree BB15, STMT00079 (before)
               [000346] DAC-G------                         *  STORE_LCL_VAR int    V10 loc4               
               [000345] --C-G------                         \--*  EQ        int   
               [000343] --C-G------                            +--*  CALLV stub int    System.IComparable`1[Program+FloatEmptyShort]:CompareTo(Program+FloatEmptyShort):int:this
               [000341] ----------- this                       |  +--*  LCL_VAR   ref    V08 loc2          (last use)
               [000342] ----------- arg1    <--@@@             |  \--*  LCL_VAR   struct<Program+FloatEmptyShort, 8> V03 arg3                <---@@@@
               [000344] -----------                            \--*  CNS_INT   int    0
Initializing arg info for 343.CALL:                                                                                                                                                           
Argument 0 ABI info: [00..08) reg a0
Argument 1 ABI info: [00..08) reg t8
Argument 2 ABI info: 2 segments
  [0] [00..04) reg f0
  [1] [06..08) reg a1
Args for call [000343] CALL after AddFinalArgsAndDetermineABIInfo:
CallArg[[000341].LCL_VAR ref (By value), 1 reg: a0, wellKnown[ThisPointer]]
CallArg[[000442].CNS_INT long (By value), 1 reg: t8, wellKnown[VirtualStubCell]]
CallArg[[000342].LCL_VAR struct (By value), 2 regs: f0 a1]
```

jakobbotsch replied:
I think you are confusing the parameters of the function being compiled and the arguments to the function call `[000343]`.

The function call `[000343]` is being passed a value as its second argument (denoted `arg1` in the JITDUMP). This value could be anything. It has no link to the second or fourth parameters in the containing method that is being compiled. It just so happens that the value passed is `V03`, i.e. we are passing the fourth parameter to the containing method as the second argument to the function call, but the value could be anything.

The ABI information for call arguments is not stored in `LclVarDsc` -- that wouldn't make sense. It's stored in `CallArg::AbiInfo` (and `CallArg::NewAbiInfo`). If you look at the information stored there, you should hopefully find what you are expecting.

LuckyXu-HF replied:
oh yes, I got confued in it, thank you!

tomeksowi commented on src/coreclr/vm/loongarch64/stubs.cpp:
```diff
@@ -1267,6 +1267,34 @@ void StubLinkerCPU::Init()
+static int GetRegister(UINT16 ofs)
+{
+    _ASSERTE(InRegister(ofs));
+    if (ofs & ShuffleEntry::FPREGMASK)
```
Nit, it will never be called with FPREGMASK, you can turn it to an assertion.

LuckyXu-HF replied:
Thanks.

tomeksowi reviewed (approved):
LGTM, some cleanup possible.

LuckyXu-HF committed:
Update stubs.cpp

jkotas commented:
@dotnet/jit-contrib Could you please sign off on the JIT change?

LuckyXu-HF committed:
clean up unused code.

LuckyXu-HF committed:
Update lclvars.cpp

Also set 'varDsc->lvIsMultiRegArg = true' in 'Compiler::lvaInitUserArgs()' when passing struct with multi regs for RiscV64.

jakobbotsch commented:
@jkotas Does the VM side look good to you?

jkotas reviewed (approved):
VM side LGTM
